### PR TITLE
Migrate validator header rules

### PR DESCRIPTION
**Pull request by fkpwolf**
[AB#268358](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/268358)

[AB#246189](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/246189)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6316)

### CODE REVIEW

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 1]
```diff
@@ -50,6 +50,8 @@ internal static class DocsValidationExtension
                             RenderedPlainText = markdownEngine.ToPlainText(headingBlock), // used for validation
                             IsVisible = MarkdigUtility.IsVisible(headingBlock),
                             IsCanonicalVersion = isCanonicalVersion,
+                            Zone = context.ZoneStack.TryPeek(out var z) ? z : null,
+                            Monikers = fileLevelMoniker.ToList(),
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:54]: this should be zoneLevelMoniker

Comment (fkpwolf) [reply src/docfx/lib/markdown/validation/DocsValidationExtension.cs]: ok

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 2]
```diff
@@ -153,6 +154,22 @@ public static void Visit(this MarkdownObject? obj, MarkdownVisitContext context,
                     context.FileStack.Pop();
                     break;
 
+                case TripleColonBlock tripleColonBlock:
+                    var target = tripleColonBlock.GetAttributes().Properties.FirstOrDefault(p => p.Key == "data-target");
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/MarkdigUtility.cs:158]: also consider the `pivot` attribute

Comment (fkpwolf) [reply src/docfx/lib/markdown/MarkdigUtility.cs:158]: ok

File: src/docfx/validation/ContentValidator.cs [hunk 3]
```diff
@@ -56,7 +58,7 @@ public void ValidateHeadings(Document file, List<ContentNode> nodes, bool isIncl
         {
             if (TryGetValidationDocumentType(file, file.Mime.Value, isIncluded, out var documentType))
             {
-                var validationContext = new ValidationContext { DocumentType = documentType };
+                var validationContext = new ValidationContext { DocumentType = documentType, FileSourceInfo = new SourceInfo(file.FilePath, 1, 1) };
```
Comment (OsmondJiang) [inline src/docfx/validation/ContentValidator.cs:61]: new SourceInfo(file.FilePath)?

Comment (fkpwolf) [reply src/docfx/validation/ContentValidator.cs]: ok

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 4]
```diff
@@ -153,6 +154,32 @@ public static void Visit(this MarkdownObject? obj, MarkdownVisitContext context,
...
+                case TripleColonBlock tripleColonBlock:
+                    string? target = null;
+                    if (tripleColonBlock.GetAttributes().Properties.Any(p => p.Key == "data-pivot"))
+                    {
+                        target = "pivot";
+                    }
+                    else
+                    {
+                        target = tripleColonBlock.GetAttributes().Properties.FirstOrDefault(p => p.Key == "data-target").Value;
+                    }
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/MarkdigUtility.cs:166]: target should have higher priority than pivot for our validation case?

Comment (fkpwolf) [reply src/docfx/lib/markdown/MarkdigUtility.cs:166]: ok

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 5]
```diff
@@ -266,7 +292,7 @@ public static bool IsVisible(this MarkdownObject markdownObject)
             {
                 var nodeVisible = node switch
                 {
-                    HtmlBlock htmlBlock => htmlBlock.Lines.Lines.Any(line => HtmlUtility.IsVisible(line.Slice.ToString())),
+                    HtmlBlock htmlBlock => HtmlUtility.IsVisible(string.Join("", htmlBlock.Lines.Lines)),
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdigUtility.cs:295]: nit: Join by `''` eats new lines:
```suggestion
                    HtmlBlock htmlBlock => HtmlUtility.IsVisible(htmlBlock.Lines.ToString()),
```

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 6]
```diff
@@ -153,6 +153,32 @@ public static void Visit(this MarkdownObject? obj, MarkdownVisitContext context,
                     context.FileStack.Pop();
                     break;
 
+                case TripleColonBlock tripleColonBlock:
+                    string? target = null;
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdigUtility.cs:157]: nit: this check is redundant, can simply
```suggestion
                    var target = tripleColonBlock.GetAttributes().Properties.FirstOrDefault(p => p.Key == "data-target")?.Value;
```

File: src/docfx/lib/markdown/MarkdigUtility.cs (hunk 5 above)
Comment (fkpwolf) [reply src/docfx/lib/markdown/MarkdigUtility.cs]: ok

File: src/docfx/lib/markdown/MarkdigUtility.cs (hunk 6 above)
Comment (fkpwolf) [reply src/docfx/lib/markdown/MarkdigUtility.cs:157]: there will be an error `c# operator ? cannot apply on type Keyvaluepair`. And we still do another checking - check if target is null and if so assign a value of `pivot`.

**Pull request merged**